### PR TITLE
Support qwen2 vl

### PR DESCRIPTION
**Pull request by drbh**
This is a work in progress PR to support qwen2-vl. Currently these changes include loading the model weights and a functioning vision model. Remaining work is to adjust the existing qwen2 model to handle multimodal requests/positional embeddings.

status:
- [X] load weights
- [X] support prefill warmup run
- [X] accept chat image request and process image
- [X] align vision model output with reference impl
- [X] correctly merge the processed image and text model
- [X] avoid any reshapes and allocations during runtime possible

remaining:
- [x] resolve remaining bug with position ids
- [x] align test output with reference
- [x] cleanup remaining todos/refactors/improvements

further
- make improvements
- improve test coverage

### CODE REVIEW

File: server/text_generation_server/layers/tensor_parallel.py [hunk 1]
```diff
@@ -144,7 +144,7 @@ def load_qkv(
             num_key_value_heads=num_key_value_heads,
         )
         if bias:
-            raise NotImplementedError("packed_qkv only implemented for baichuan")
+            bias = weights.get_tensor(f"{prefix}.bias")
```
Comment (Narsil) [inline server/text_generation_server/layers/tensor_parallel.py:147]: This is wrong, no ?

We get the whole bias on Row Parallel, for column, you need to take the actual slice, which for qkv you need to follow the same layout as the weights I think (except on dim=0 instead of dim=1)

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 2]
```diff
@@ -122,7 +130,28 @@ def forward(
         query = query.view(-1, self.num_heads, self.head_size)
         kv = kv.view(-1, 2, self.num_key_value_heads, self.head_size)
 
-        self.rotary_emb(query, torch.select(kv, dim=1, index=0), cos, sin)
+        # TODO: correctly handle the multimodal case
+        if False:
+            self.rotary_emb(query, torch.select(kv, dim=1, index=0), cos, sin)
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py:135]: This should be correct. Ignore tiny differences there, this code is exactly what you have underneath (and much more efficient).

I read transformers comment on this, it seems from what I'm reading that they are just applying part of the tensors there, so a regular slicing should do the work.

The problem with the other part, is that our cos, sin are layed out differently than theirs, so you're gonna have issues keeping transformers code and merge it with our own.

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 3]
```diff
@@ -306,12 +335,24 @@ def forward(
         max_s: int,
         true_max_s: int,
         prefill_cache_indices: Optional[torch.Tensor],
+        inputs_embeds: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
     ) -> torch.Tensor:
-        hidden_states = self.embed_tokens(input_ids)
+
+        # if inputs_embeds are supplied from an external model (vision model) then avoid embedding input_ids
+        if inputs_embeds is not None:
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py:343]: No, remove this.

The way it's done here, is that we life input_ids to the parent class, and this always takes `input_embeds`. Makes signatures much cleaner. (It's already done for llama this way if you want to check)

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 4]
```diff
@@ -306,12 +335,24 @@ def forward(
         max_s: int,
         true_max_s: int,
         prefill_cache_indices: Optional[torch.Tensor],
+        inputs_embeds: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py:339]: No attention_mask

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+                    (input_ids == self.image_token_id)
+                    .unsqueeze(-1)
+                    .expand_as(inputs_embeds)
+                    .to(inputs_embeds.device)
+                )
+                image_embeds = image_embeds.to(
+                    inputs_embeds.device, inputs_embeds.dtype
+                )
+                # input embeddings are masked with image embeddings
+                inputs_embeds = inputs_embeds.masked_scatter(image_mask, image_embeds)
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/qwen2_vl.py:415]: Why does

```python
inputs_embeds[input_ids == self.image_token_id] = image_embeds
```

Doesn't work ?

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+        image_sizes: Optional[torch.LongTensor] = None,
+        adapter_data: Optional[torch.Tensor] = None,
+        cross_attention_states: Optional[torch.Tensor] = None,
+        image_indices=None,
+    ):
+        # make an attention_mask that is (batch_size, sequence_length)
+        attention_mask = torch.ones_like(
+            input_ids, dtype=torch.bool, device=input_ids.device
+        )
+        inputs_embeds = self.text_model.embed_tokens(input_ids)
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/qwen2_vl.py:399]: When you lift this, this will be gone from the `text_model` and be here directly.

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+                3,
+                input_ids.shape[0],
+                input_ids.shape[1],
+                dtype=input_ids.dtype,
+                device=input_ids.device,
+            )
+            image_index, video_index = 0, 0
+
+            for i, input_ids in enumerate(total_input_ids):
+                if attention_mask is not None:
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/qwen2_vl.py:431]: This whole thing is extremely poor code (lots of loops, lots of CPU/GPU back&forth).

I think ditching it altogether will be easier than trying to adapt it.

File: server/text_generation_server/layers/tensor_parallel.py (hunk 1 above)
Comment (drbh) [reply server/text_generation_server/layers/tensor_parallel.py]: after comparing with transformers it seems like `weights.get_tensor(f"{prefix}.bias")` and `weights.get_sharded(f"{prefix}.qkv.bias", dim=0)` return the exact bias as the one in the reference.

I've reverted the change within `tensor_parallel.py::load_qkv` in favor of setting the bias after creating the linear in `qwen_vl.py` via `weights.get_sharded(f"{prefix}.qkv.bias", dim=0)`

for reference:
```python
self.qkv = TensorParallelColumnLinear.load_qkv(
    config,
    prefix=f"{prefix}.qkv",
    weights=weights,
    bias=False,
    num_heads=self.num_heads,
    num_key_value_heads=self.num_heads,
)
self.qkv.linear.bias = weights.get_sharded(f"{prefix}.qkv.bias", dim=0)
```

this hopefully makes the qkv loading a bit more clear

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py (hunk 3 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py]: yes agreed this is much cleaner. I've updated the classes in the latest commit to follow the same pattern as llama

File: server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py (hunk 4 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/flash_qwen2_modeling.py]: removed in latest commit

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py (hunk 5 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/qwen2_vl.py]: ha yea thats a much better way to write this! I've included this change along with a overall better rewrite of this logic in the latest commit.

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py (hunk 6 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/qwen2_vl.py]: updated in latest commit

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py (hunk 7 above)
Comment (drbh) [reply server/text_generation_server/models/custom_modeling/qwen2_vl.py]: yea agreed, in the latest commit i've moved this logic into a function `get_position_ids` and rewritten a more simple version that avoids most of the gpu/cpu copies. I'll revisit later to see if I can simplify further (avoid the loop) but the changes may provide a bit of a performance and readability improvement.

File: server/text_generation_server/models/vlm_causal_lm.py [hunk 8]
```diff
@@ -343,6 +363,16 @@ def forward(
...
 
+        if hasattr(self.model, "get_position_ids"):
+            if position_ids.shape[0] != 1:
+                position_ids = self.model.get_position_ids(
+                    input_ids.unsqueeze(0), batch.image_grid_thw
+                )
+                batch.position_ids = position_ids[0, 0, :]
+            else:
+                position_ids = position_ids.repeat(3, 1, 1).clone()
+                batch.position_ids = position_ids[0, 0, :]
```
Comment (Narsil) [inline server/text_generation_server/models/vlm_causal_lm.py:374]: This seems very wrong, no ?

File: server/text_generation_server/models/vlm_causal_lm.py [hunk 9]
```diff
@@ -343,6 +363,16 @@ def forward(
             max_s = batch.max_current_length
             lm_head_indices = batch.prefill_head_indices
 
+        if hasattr(self.model, "get_position_ids"):
+            if position_ids.shape[0] != 1:
+                position_ids = self.model.get_position_ids(
+                    input_ids.unsqueeze(0), batch.image_grid_thw
+                )
+                batch.position_ids = position_ids[0, 0, :]
```
Comment (Narsil) [inline server/text_generation_server/models/vlm_causal_lm.py:371]: Why create so many position ids, just to discard them ?

File: server/text_generation_server/models/custom_modeling/qwen2_vl.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+            prefix="lm_head", weights=weights, config=config, bias=False
+        )
+        self.norm = FastRMSNorm.load(
+            prefix="model.norm",
+            weights=weights,
+            eps=config.rms_norm_eps,
+        )
+        self.device = weights.device
+
+    def get_position_ids(
```
Comment (Narsil) [inline server/text_generation_server/models/custom_modeling/qwen2_vl.py:377]: Seems overly complex and bloated.

Let's keep this if it's working, but it's definitely fixable I think

File: server/text_generation_server/models/vlm_causal_lm.py [hunk 11]
```diff
@@ -343,6 +363,16 @@ def forward(
             max_s = batch.max_current_length
             lm_head_indices = batch.prefill_head_indices
 
+        if hasattr(self.model, "get_position_ids"):
+            if position_ids.shape[0] != 1:
+                position_ids = self.model.get_position_ids(
+                    input_ids.unsqueeze(0), batch.image_grid_thw
```
Comment (Narsil) [inline server/text_generation_server/models/vlm_causal_lm.py:369]: No unsqueeze, fix the function.

**Review by Narsil — APPROVED** [review]

**Comment by drbh** [comment]
merging to enable qwen2-vl

Will follow up with an improvement PR soon. specifically:
- [ ] improve decode
- [ ] improve vision head
- [ ] improve batch to handle multi dimensional position ids
- [ ] remove complex position logic if possible
